clippy_lints/src/functions/result.rs

```diff
@@ -0,0 +1,152 @@
+use rustc_errors::Diagnostic;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty;
+use rustc_span::{sym, Span};
+use rustc_typeck::hir_ty_to_ty;
+
+use clippy_utils::diagnostics::{span_lint_and_help, span_lint_and_then};
+use clippy_utils::trait_ref_of_method;
+use clippy_utils::ty::is_type_diagnostic_item;
+
+use super::{RESULT_LARGE_ERR, RESULT_UNIT_ERR};
+
+/// comes up with an "at least" guesstimate for the type's size, not taking into
+/// account the layout of type parameters.
+/// For example, a `((u32, u32, T), T, [u8; 4])` will be 12, which is only
+/// correct if `T` is a zst *and* doesn't have alignment...
+fn approx_ty_size<'tcx>(cx: &LateContext<'tcx>, ty: ty::Ty<'tcx>) -> u64 {
+    use rustc_middle::ty::layout::LayoutOf;
+    match (cx.layout_of(ty).map(|layout| layout.size.bytes()), ty.kind()) {
+        (Ok(size), _) => size,
+        (Err(_), ty::Tuple(list)) => list.as_substs().types().map(|t| approx_ty_size(cx, t)).sum(),
+        (Err(_), ty::Adt(def, _)) if def.is_struct() => def
+            .variants()
+            .iter()
+            .map(|v| {
+                v.fields
+                    .iter()
+                    .map(|field| approx_ty_size(cx, cx.tcx.type_of(field.did)))
+                    .sum::<u64>()
+            })
+            .sum(),
+        (Err(_), ty::Adt(def, _)) if def.is_enum() => def
+            .variants()
+            .iter()
+            .map(|v| {
+                v.fields
+                    .iter()
+                    .map(|field| approx_ty_size(cx, cx.tcx.type_of(field.did)))
+                    .sum::<u64>()
+            })
+            .max()
+            .unwrap_or_default(),
+        (Err(_), ty::Adt(def, _)) if def.is_union() => def
+            .variants()
+            .iter()
+            .map(|v| {
+                v.fields
+                    .iter()
+                    .map(|field| approx_ty_size(cx, cx.tcx.type_of(field.did)))
+                    .max()
+                    .unwrap_or_default()
+            })
+            .max()
+            .unwrap_or_default(),
+        (Err(_), _) => 0,
+    }
+}
+
+/// The type of the `Err`-variant in a `std::result::Result` returned by the
+/// given `FnDecl`
+fn result_err_ty<'tcx>(
+    cx: &LateContext<'tcx>,
+    decl: &hir::FnDecl<'tcx>,
+    item_span: Span,
+) -> Option<(&'tcx hir::Ty<'tcx>, ty::Ty<'tcx>)> {
+    if !in_external_macro(cx.sess(), item_span)
+    && let hir::FnRetTy::Return(hir_ty) = decl.output
+    && let ty = hir_ty_to_ty(cx.tcx, hir_ty)
+    && is_type_diagnostic_item(cx, ty, sym::Result)
+    && let ty::Adt(_, substs) = ty.kind() {
```

Comment:
Some formatting as rustfmt doesn't do let chains yet:

```suggestion
        && let hir::FnRetTy::Return(hir_ty) = decl.output
        && let ty = hir_ty_to_ty(cx.tcx, hir_ty)
        && is_type_diagnostic_item(cx, ty, sym::Result)
        && let ty::Adt(_, substs) = ty.kind()
    {
```